File: product-microservice/api/console_api.go
package api

import (
	"product/dto"
	"product/dto/filter"
	"product/mapper"
	"product/service"
	"strconv"

	"net/http"

	"github.com/gin-gonic/gin"
)


type ConsoleAPI struct {
	IConsoleService service.IConsoleService
}

func NewConsoleAPI(consoleService service.IConsoleService) ConsoleAPI {
	return ConsoleAPI{IConsoleService: consoleService}
}

func (consoleApi *ConsoleAPI) GetAll(c *gin.Context) {
	page, err := strconv.Atoi(c.Query("page"))
	pageSize, err := strconv.Atoi(c.Query("pageSize"))
    if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
    }
	
	consoles := consoleApi.IConsoleService.GetAll(page, pageSize)
	c.JSON(http.StatusOK, mapper.ToConsoleDTOs(consoles))
}

func (consoleApi *ConsoleAPI) GetNumberOfRecords(c *gin.Context) {
	numberOfRecords := consoleApi.IConsoleService.GetNumberOfRecords()
	c.JSON(http.StatusOK, numberOfRecords)
}

func (consoleApi *ConsoleAPI) GetByID(c *gin.Context) {
	id, err := strconv.Atoi(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}
	
	console, err := consoleApi.IConsoleService.GetById(id)
	
	if err == nil {
		c.JSON(http.StatusOK, mapper.ToConsoleDTO(console))
	} else {
		c.JSON(http.StatusBadRequest, err.Error())
	}
}

func (consoleApi *ConsoleAPI) SearchByName(c *gin.Context) {
	page, err := strconv.Atoi(c.Query("page"))
	pageSize, err := strconv.Atoi(c.Query("pageSize"))
    if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
    }

	consoles, err := consoleApi.IConsoleService.SearchByName(page, pageSize, c.Query("name"))
	
	if err == nil {
		c.JSON(http.StatusOK, mapper.ToConsoleDTOs(consoles))
	} else {
		c.JSON(http.StatusBadRequest, err.Error())
	}
}

func (consoleApi *ConsoleAPI) GetNumberOfRecordsSearch(c *gin.Context) {
	numberOfRecords := consoleApi.IConsoleService.GetNumberOfRecordsSearch(c.Query("name"))
	c.JSON(http.StatusOK, numberOfRecords)
}

func (consoleApi *ConsoleAPI) Filter(c *gin.Context) {
	page, err := strconv.Atoi(c.Query("page"))
	pageSize, err := strconv.Atoi(c.Query("pageSize"))
    if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
    }

	var filter filter.ConsoleFilter
	err = c.BindJSON(&filter)
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}

	consoles, err := consoleApi.IConsoleService.Filter(page, pageSize, filter)
	
	if err == nil {
		c.JSON(http.StatusOK, mapper.ToConsoleDTOs(consoles))
	} else {
		c.JSON(http.StatusBadRequest, err.Error())
	}
}

func (consoleApi *ConsoleAPI) GetNumberOfRecordsFilter(c *gin.Context) {
	var filter filter.ConsoleFilter
	err := c.BindJSON(&filter)
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}

	numberOfRecords := consoleApi.IConsoleService.GetNumberOfRecordsFilter(filter)
	c.JSON(http.StatusOK, numberOfRecords)
}

func (consoleApi *ConsoleAPI) GetPlatforms(c *gin.Context) {
	platforms := consoleApi.IConsoleService.GetPlatforms()
	c.JSON(http.StatusOK, platforms)
}

func (consoleApi *ConsoleAPI) Create(c *gin.Context) {
	var consoleDTO dto.ConsoleDTO
	err := c.BindJSON(&consoleDTO)
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}

	console := mapper.ToConsole(consoleDTO)
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}

	msg := consoleApi.IConsoleService.Create(console)

	if msg == "" {
		c.JSON(http.StatusOK, "Console added successfully.")
	} else {
		c.JSON(http.StatusBadRequest, msg)
	}
}

func (consoleApi *ConsoleAPI) Update(c *gin.Context) {
	var consoleDTO dto.ConsoleDTO
	err := c.BindJSON(&consoleDTO)
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}

	msg := consoleApi.IConsoleService.Update(consoleDTO)

	if msg == "" {
		c.JSON(http.StatusOK, "Console updated successfully.")
	} else  {
		c.JSON(http.StatusBadRequest, msg)
	} 
}

func (consoleApi *ConsoleAPI) Delete(c *gin.Context) {
	id, err := strconv.Atoi(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}
	
	error := consoleApi.IConsoleService.Delete(id)

	if error == nil {
		c.JSON(http.StatusOK, "Console deleted successfully.")
	} else  {
		c.JSON(http.StatusBadRequest, error.Error())
	}
}

File: api-gateway/api/product_microservice/console_api.py
import requests
from flask import request, jsonify
from utils import token_utils, role
from utils.routes.product_microservice import console_api_routes


def get_all():
    args = request.args.to_dict()
    page = args.get("page")
    page_size = args.get("pageSize")
    r = requests.get(console_api_routes.BASE + console_api_routes.API +
                        console_api_routes.GET_ALL
                        + "?page={}&pageSize={}".format(page, page_size))
    resp = jsonify(r.json())
    resp.status_code = r.status_code
    return resp


def get_number_of_records():
    r = requests.get(console_api_routes.BASE + console_api_routes.API +
                        console_api_routes.GET_NUMBER_OF_RECORDS)
    resp = jsonify(r.json())
    resp.status_code = r.status_code
    return resp


def get_by_id(id):
    r = requests.get(console_api_routes.BASE + console_api_routes.API +
                        console_api_routes.GET_BY_ID + "/{}".format(id))
    resp = jsonify(r.json())
    resp.status_code = r.status_code
    return resp


def search_by_name():
    args = request.args.to_dict()
    page = args.get("page")
    page_size = args.get("pageSize")
    name = args.get("name")
    r = requests.get(console_api_routes.BASE + console_api_routes.API +
                        console_api_routes.SEARCH_BY_NAME
                        + "?page={}&pageSize={}&name={}".format(page, page_size, name))
    resp = jsonify(r.json())
    resp.status_code = r.status_code
    return resp


def get_number_of_records_search():
    args = request.args.to_dict()
    name = args.get("name")
    r = requests.get(console_api_routes.BASE + console_api_routes.API +
                        console_api_routes.GET_NUMBER_OF_RECORDS_SEARCH + "?name={}".format(name))
    resp = jsonify(r.json())
    resp.status_code = r.status_code
    return resp


def filter():
    args = request.args.to_dict()
    page = args.get("page")
    page_size = args.get("pageSize")
    data = request.json
    r = requests.post(console_api_routes.BASE + console_api_routes.API +
                         console_api_routes.FILTER +
                         "?page={}&pageSize={}".format(page, page_size), json=data)
    resp = jsonify(r.json())
    resp.status_code = r.status_code
    return resp


def get_number_of_records_filter():
    data = request.json
    r = requests.post(console_api_routes.BASE + console_api_routes.API +
                         console_api_routes.GET_NUMBER_OF_RECORDS_FILTER, json=data)
    resp = jsonify(r.json())
    resp.status_code = r.status_code
    return resp


def get_platforms():
    r = requests.get(console_api_routes.BASE + console_api_routes.API +
                        console_api_routes.GET_PLATFORMS)
    resp = jsonify(r.json())
    resp.status_code = r.status_code
    return resp


@token_utils.authorization_required(roles=[role.ROLE_EMPLOYEE])
def create():
    data = request.json
    headers = request.headers
    r = requests.post(console_api_routes.BASE + console_api_routes.API +
                         console_api_routes.CREATE, json=data, headers=headers)
    resp = jsonify(r.json())
    resp.status_code = r.status_code
    return resp


@token_utils.authorization_required(roles=[role.ROLE_EMPLOYEE])
def update():
    data = request.json
    headers = request.headers
    r = requests.put(console_api_routes.BASE + console_api_routes.API +
                        console_api_routes.UPDATE, json=data, headers=headers)
    resp = jsonify(r.json())
    resp.status_code = r.status_code
    return resp


@token_utils.authorization_required(roles=[role.ROLE_EMPLOYEE])
def delete(id):
    headers = request.headers
    r = requests.delete(console_api_routes.BASE + console_api_routes.API +
                           console_api_routes.DELETE + "/{}".format(id), headers=headers)
    resp = jsonify(r.json())
    resp.status_code = r.status_code
    return resp
